src/shogun/distance/EuclideanDistance.cpp

```diff
@@ -177,3 +179,38 @@ float64_t CEuclideanDistance::distance_upper_bounded(int32_t idx_a, int32_t idx_
 
 	return result;
 }
+
+SGMatrix<float64_t> CEuclideanDistance::get_distance_matrix()
+{
+	REQUIRE(lhs->get_feature_class()==C_DENSE,
+		"Left hand side (was %s) has to be CDenseFeatures instance!\n", lhs->get_name());
+	REQUIRE(rhs->get_feature_class()==C_DENSE,
+		"Right hand side (was %s) has to be CDenseFeatures instance!\n", rhs->get_name());
+
+	REQUIRE(lhs->get_feature_type()==F_DREAL,
+		"Left hand side (was %s) has to be of double type!\n", lhs->get_name());
+	REQUIRE(rhs->get_feature_type()==F_DREAL,
+		"Right hand side (was %s) has to be double type!\n", rhs->get_name());
```

Comment:
Hi @Saurabh7. The method `get_distance_matrix` doesn't have to be a template. It should always return a `SGMatrix<float64_t>` (even for sparse features). But internally it should be able to handle both `CDenseFeatures<float64_t>` and `CDenseFeatures<float32_t>`, along with other feature types that supports `FP_DOT`. You can have a template private method for that which handles both the cases - you do the Eigen3 stuffs inside that. Fallback works with `dot` method in the features. Sounds reasonable?

Another thing I missed - the present implementation creates two new feature matrices every time you call this (Please check [`CDenseFeatures::get_feature_matrix()`](https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/features/DenseFeatures.cpp#L273)). So, I think we should have two of these methods

1. One that works without subsets (fail with `SG_ERROR` if subset is present, say that the user should use the other variant if they subsets). It should use [`CDenseFeatures::get_feature_matrix(int32_t&, int32_t&)`](https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/features/DenseFeatures.cpp#L306), then initialize the Eigen3 maps with that. No mem alloc/copy, no-ref-counting business, faster.
2. One that works with subsets. This may give a `SG_INFO` when subsets are **not** present - saying that the users are better off using the other variant. This requires mem alloc/copy, ref-counting for SGMatrix but respects subsets.

We can do this by either having the method as `virtual SGMatrix<float64_t> get_distance_matrix(bool subset_present=false)` or having two different interfaces `get_distance_matrix()` and `get_distance_matrix_with_subsets()`. Thoughts on this?

When we're done with the implementation, we should also have unit tests for all the feature types for which `EuclideanDistance` can be initialized. This way, if something doesn't work, we'll catch it.

Let me know what you think. @karlnapf may have some comments.